### PR TITLE
Geospatial search

### PR DESCRIPTION
**Pull request by qqmyers**
**What this PR does / why we need it**: Adds bounding box indexing in preparation for providing geosearch or other features

**Which issue(s) this PR closes**:

Closes #

**Special notes for your reviewer**:

**Suggestions on how to test this**:

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

**Is there a release notes update needed for this change?**:

**Additional documentation**: design doc: https://docs.google.com/document/d/1aQCIM6wbAdgMNLTVSUhvuHQKzWDAvQ5XoYGbfint3P4/edit?usp=sharing

### CODE REVIEW

**Comment by qqmyers** [comment]
Per discussion at the DCM that the improved geospatial indexing could be useful independent of adding further geospatial search/display capabilities, I'm moving this PR from draft to review-ready. Not sure what the Rocky shellspec issue is - probably worth re-running the build to see if there's any real issue.

File: conf/solr/8.11.1/schema.xml [hunk 1]
```diff
@@ -647,6 +647,9 @@
     <copyField source="westLongitude" dest="_text_" maxChars="3000"/>
     <!-- SCHEMA-COPY-FIELDS::END -->
```
Comment (pdurbin) [inline conf/solr/8.11.1/schema.xml:648]: I feel like we need some documentation. Maybe something in the Search API page?

**Review by pdurbin — COMMENTED** [review]
@qqmyers this looks promising! Can you please merge from develop and add a release note? What about docs? I feel like we should add something...

**Comment by qqmyers** [comment]
I'm not sure it is possible to make an IT test yet. The solr query that works is as shown:
image

In the UI, I don't think we can submit filterqueries and with the API, you can submit the filter query part but I haven't seen how to send the pt and d parameters yet. If that is possible, an IT test would be straight forward. 

Otherwise, to test one could use the solr console (dropping the jetty.host-localhost recommended for our solr install to make it accessible, and possibly doing an `aws ec2 authorize-security-group-ingress --group-id <id> --protocol tcp --port 8983 --cidr <your ip>/32` to let a browser access the console.)

File: conf/solr/8.11.1/schema.xml [hunk 2]
```diff
@@ -647,6 +647,9 @@
     <copyField source="westLongitude" dest="_text_" maxChars="3000"/>
```
[the remark this answers is not in the document]
Comment (qqmyers) [reply conf/solr/8.11.1/schema.xml:647]: added

File: conf/solr/8.11.1/schema.xml (hunk 1 above)
Comment (qqmyers) [reply conf/solr/8.11.1/schema.xml:648]: Put something short in. Feel free to modify/add.

File: conf/solr/8.11.1/schema.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pdurbin) [reply conf/solr/8.11.1/schema.xml:647]: Thanks!

File: conf/solr/8.11.1/schema.xml (hunk 1 above)
Comment (pdurbin) [reply conf/solr/8.11.1/schema.xml:648]: Looks like a good start! Thanks much.

**Comment by pdurbin** [comment]
I'm on 202438a. When I run DataversesIT.testImportDDI it fails with this:

INFO: {"status":"ERROR","message":"Command [DatasetCreate dataset:3] failed: Exception thrown from bean: javax.ejb.EJBTransactionRolledbackException: Exception thrown from bean: org.apache.solr.client.solrj.impl.HttpSolrClient$RemoteSolrException: Error from server at http://localhost:8983/solr/collection1: Exception writing document id dataset_3_draft to the index; possible analysis error: DocValuesField \"solr_bboxtype__minX\" appears more than once in this document (only one value is allowed per field)"}

@qqmyers can you please take a look? I did update to the latest schema.xml you provided. I'll merge the latest from develop into this branch to trigger a Jenkins run. Thanks.

**Comment by pdurbin** [comment]
In c7c16d4 I updated the Search API two allow two new geo-related parameters: `geo_point` and `geo_radius`:

Screen Shot 2022-10-24 at 11 52 26 AM

I created a dataset using lat/logs from a box I drew and copied from https://boundingbox.klokantech.com

`westlimit=-71.187346; southlimit=42.33661; eastlimit=-71.043056; northlimit=42.409599`

Screen Shot 2022-10-24 at 11 30 51 AM

If I use a radius of 1 km, I can't find the dataset:

```
$ curl "http://localhost:8080/api/search?q=*&geo_point=42.3,-71.1&geo_radius=1"
```

`{"status":"OK","data":{"q":"*","total_count":0,"start":0,"spelling_alternatives":{},"items":[],"count_in_response":0}}`

If I use a radius of 4 km, I get a hit:

```
$ curl "http://localhost:8080/api/search?q=*&geo_point=42.3,-71.1&geo_radius=4"
```

`{"status":"OK","data":{"q":"*","total_count":1,"start":0,"spelling_alternatives":{},"items":[{"name":"Dataverse HQ","type":"dataset","url":"https://doi.org/10.5072/FK2/MKEOMD"...`

Right now this is API-only. I'm not sure how much work it would be to add it to the GUI.

Right now both `geo_point` and `geo_radius` are required to trigger geospatial search. Instead should I set `geo_radius` to a sane default (1 km? 10 km? 100 km?) and only make `geo_point` required to trigger geospatial search?

Do you like the names of these new parameters? My goal was to make them user friendly. `geo_point` maps to `pt` in Solr. `geo_radius` maps to `d` in Solr. You can read about these parameters at https://solr.apache.org/guide/8_11/spatial-search.html and here's a screenshot from those docs:

Screen Shot 2022-10-24 at 12 03 42 PM

**Comment by pdurbin** [comment]
I added a geospatial search test in 7272de6.

@qqmyers I just noticed that if I comment out `northLongitude`...

```
//.add("northLongitude",
//        Json.createObjectBuilder()
//                .add("value", "42.409599")
//                .add("typeClass", "primitive")
//                .add("multiple", false)
//                .add("typeName", "northLongitude")
//)
```
... it gets populated in `solr_srpt` and `solr_bboxtype` with the value from `southLongitude`:

        "solr_srpt":["ENVELOPE(-71.187346,-71.043056,42.33661,42.33661)"],
        "solr_bboxtype":["ENVELOPE(-71.187346,-71.043056,42.33661,42.33661)"],
        "westLongitude":["-71.187346"],
        "southLongitude":["42.33661"],
        "eastLongitude":["-71.043056"],

What do you think? Bug or feature? 😄 

p.s. Yes, I know it should be `northLatitude` and `southLatitude`. There's a bug tracking this:

- #5645

**Comment by qqmyers** [comment]
I think the indexing code assumes you're talking about a point (or line) if you leave one coord out so it just replicates the value for the other extreme. So ~feature.

**Comment by pdurbin** [comment]
@qqmyers ok, but I did go ahead and fix a bug when there was only one coordinate: e5187b2af4

Next up: names. I haven't heard any objections to `geo_point` and `geo_radius` so I'm planning on sticking with them. Unless we want `spatial_point` and `spatial_radius`? I'm pretty sure we're limiting ourselves to Earth, though, so I think "geo" is fine (and shorter!). 🌎 

On the Solr field side we currently have `solr_bboxtype` and `solr_srpt`. None of our existing fields start with `solr_` so I'm thinking of lopping that off.

Any reason not to shorted to just `bbox`? Our Solr fields usually don't end with `type` and just `bbox` is what's shown in the docs: https://solr.apache.org/guide/8_11/spatial-search.html#bboxfield

I don't know what to say about RPT. SpatialRecursivePrefixTreeFieldType is a mouthful. I'm not in love with `solr_srpt` but the only example I see in the docs is `location_rpt`. Perhaps we should go with that.

**Comment by pdurbin** [comment]
I added some error checking and updated the docs based on how the code is now.

I talked to @atrisovic about this feature and she immediately asked "Can you sort by distance?" Good question! 😄  I haven't looked into this but maybe? See https://solr.apache.org/guide/8_11/spatial-search.html#distance-sorting-or-boosting-function-queries for `sort=geodist() asc`. I'd need more data to test with.

I'm also ok with this being API only for now. That's how I wrote it up in the User Guide: https://dataverse-guide--8239.org.readthedocs.build/en/8239/user/find-use-data.html#geospatial-search

I'm ready for review. We could keep hacking but there's lots of other stuff to do as well. 😄 I think this is an excellent toehold into geospatial search!

**Comment by scolapasta** [comment]
Moved back to "This Sprint" so that @pdurbin's recent changes can be reviewed. (@pdurbin reviewed the initial PR and was fine with the changes Jim made.)

File: src/test/java/edu/harvard/iq/dataverse/api/SearchIT.java [hunk 3]
```diff
@@ -1222,18 +1223,47 @@ public void testGeospatialSearch() {
...
+        junkPoint.prettyPrint();
+        junkPoint.then().assertThat()
+                .statusCode(BAD_REQUEST.getStatusCode())
+                .body("message", CoreMatchers.equalTo("Must contain a single comma to separate latitude and longitude."));
+
+        Response junkRadius = UtilIT.search("*", null, "&geo_point=40,60&geo_radius=junk");
+        junkRadius.prettyPrint();
+        junkRadius.then().assertThat()
+                .statusCode(BAD_REQUEST.getStatusCode())
+                .body("message", CoreMatchers.equalTo("Non-number radius supplied."));
```
Comment (sekmiller) [inline src/test/java/edu/harvard/iq/dataverse/api/SearchIT.java:1267]: Should we add a test that's actually a valid lat, long and radius?
or a valid lat with a non-numeric long, etc.?

File: src/main/java/edu/harvard/iq/dataverse/api/Search.java [hunk 4]
```diff
@@ -126,6 +126,14 @@ public Response search(
...
+                    return error(Response.Status.BAD_REQUEST, "If you supply geo_point you must also supply geo_radius.");
+                }
+
+                if (geoRadius != null && geoPoint == null) {
+                    return error(Response.Status.BAD_REQUEST, "If you supply geo_radius you must also supply geo_point.");
+                }
+
             } catch (Exception ex) {
                 return error(Response.Status.BAD_REQUEST, ex.getLocalizedMessage());
             }
```
Comment (sekmiller) [inline src/main/java/edu/harvard/iq/dataverse/api/Search.java:139]: Do we want to catch a nonnumeric exception here and give more feedback?

**Review by sekmiller — COMMENTED** [review]
Mainly looked at Phil's most recent changes. Looks good, just a couple of suggestions/questions about testing and error handling.

File: src/main/java/edu/harvard/iq/dataverse/api/Search.java (hunk 4 above)
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/api/Search.java:139]: Well, getGeoPoint and getGeoRadius will already throw exceptions with feedback about non-numeric values.

For example, if the user does this:

    /api/search?q=*&geo_point=41.9580775,-70.6621063&geo_radius=junk

They'll get this error:

    {"status":"ERROR","message":"Non-number radius supplied."}

That said, I may not be following the use cases, the concerns. I'm happy to add more tests for these.

File: src/test/java/edu/harvard/iq/dataverse/api/SearchIT.java (hunk 3 above)
Comment (pdurbin) [reply src/test/java/edu/harvard/iq/dataverse/api/SearchIT.java:1274]: In 364e347 I added a test for if lat/long is too big. It turns out Solr provides a pretty useful error, which was already being bubbled up.

I also took a swing at sending a huge value like "99999999999999999999999999999999" to `geo_radius` but Solr didn't give an error. I dunno, we already test that this number is positive. I say we let the use send bigger and bigger numbers (these are in kilometers). Eventually, they'll figure out they just aren't going to get a geospatial hit. 😄

**Review by pdurbin — COMMENTED** [review]
I'm replying to review feedback.

File: src/main/java/edu/harvard/iq/dataverse/search/IndexServiceBean.java [hunk 5]
```diff
@@ -947,6 +948,47 @@ public SolrInputDocuments toSolrDocs(IndexableDataset indexableDataset, Set<Long
...
+                            } else if (westLon == null) {
+                                westLon = eastLon;
+                            }
+                            if (northLat == null) {
+                                northLat = southLat;
+                            } else if (southLat == null) {
+                                southLat = northLat;
+                            }
+                            //W, E, N, S
+                            solrInputDocument.addField("solr_srpt", "ENVELOPE(" + westLon + "," + eastLon + "," + northLat + "," + southLat + ")");
```
[the remark this answers is not in the document]
Comment (pdurbin) [reply src/main/java/edu/harvard/iq/dataverse/search/IndexServiceBean.java]: Whoops, this was a stray comment from before. All fixed not. Resolving.

**Review by sekmiller — APPROVED** [review]
Looks good, thanks.

**Comment by mreekie** [comment]
grooming
- Looking at the NetCDF deliverables.
- This looks like it's a related issue that is related to active work, but was done before we were trying to tag each issue
- Adding it to the backlog as cleared, and tagging it.

**Pull request merged**